File: python/bob/bob.py
def response(hey_bob):
    question = hey_bob.rstrip().endswith('?')
    yell = any(c.isupper() for c in hey_bob) \
        and not any(c.islower() for c in hey_bob)
    nothing = not hey_bob.strip()

    if question and not yell:
        return "Sure."
    elif not question and yell:
        return "Whoa, chill out!"
    elif question and yell:
        return "Calm down, I know what I'm doing!"
    elif nothing:
        return "Fine. Be that way!"
    else:
        return "Whatever."

File: haskell/bob/src/Bob.hs
module Bob
  ( responseFor
  ) where

import           Data.Char                      ( isLower
                                                , isSpace
                                                , isUpper
                                                )
import           Data.String.Utils              ( endswith
                                                , rstrip
                                                )

responseFor :: String -> String
responseFor xs | isQuestion && not isYell = "Sure."
               | not isQuestion && isYell = "Whoa, chill out!"
               | isQuestion && isYell     = "Calm down, I know what I'm doing!"
               | isSayNothing             = "Fine. Be that way!"
               | otherwise                = "Whatever."
 where
  isQuestion   = endswith "?" (rstrip xs)
  isYell       = any isUpper xs && not (any isLower xs)
  isSayNothing = all isSpace xs
